modules/cart/src/EventSubscriber/OrderEventSubscriber.php

```diff
@@ -46,4 +47,17 @@ public function finalizeCart(WorkflowTransitionEvent $event) {
     }
   }
 
+  /**
+   * Finalizes the cart when the order is cancelled.
+   *
+   * @param \Drupal\state_machine\Event\WorkflowTransitionEvent $event
+   *   The workflow transition event.
+   */
+  public function cancelCart(WorkflowTransitionEvent $event) {
+    $order = $event->getEntity();
+    if ($order->cart->value == TRUE) {
+      $this->cartProvider->finalizeCart($order, FALSE);
```

Comment:
This is wrong, because finalizeCart adds the cart ID to the completed carts list in the session.
I guess CartProvider needs a cancelCart as well.